File: unity_path/src/joy_cmd_vel.cpp
#include "unity_path/joy_cmd_vel.h" 

using namespace unity_path;
using namespace std;


        
      
JoyCmdvel::JoyCmdvel(ros::NodeHandle& nh, ros::NodeHandle& nh_local) : nh_(nh), nh_local_(nh_local) {
    
    
    nh_local_.param<string>("cmd_vel_topic",cmd_vel_topic_,"/robot/robotnik_base_control/cmd_vel");
    
    
        
   
    
    yaw_rate_sub_ = nh_.subscribe("/unity/yaw_rate", 100, &JoyCmdvel::YawRateCallback,this);
    forward_vel_sub_ = nh_.subscribe("/unity/forward_vel", 100, &JoyCmdvel::ForwardVelCallback,this);

 
    cmd_vel_pub_ = nh_.advertise<geometry_msgs::Twist>(cmd_vel_topic_, 10);
    PublishCmdVel();
}



void JoyCmdvel::YawRateCallback (const std_msgs::Float64::ConstPtr& yaw_msg)
{

    angular_vel = yaw_msg->data;
     
}

void JoyCmdvel::ForwardVelCallback (const std_msgs::Float64::ConstPtr& vel_msg)
{

    linear_vel = vel_msg->data;

     
}


void JoyCmdvel::PublishCmdVel()
{
   ros::Rate r(20.0);
 
   cmd_vel_msg.linear.x = 0.0;
   cmd_vel_msg.angular.z = 0.0;

   
   while(ros::ok())
      {
        
        cmd_vel_msg.linear.x = linear_vel;
        cmd_vel_msg.angular.z = -angular_vel;

        cmd_vel_pub_.publish(cmd_vel_msg);     

  
      
        ros::spinOnce();
        r.sleep();
      }
     
                      
}

JoyCmdvel::~JoyCmdvel()
 {

 }

File: unity_path/scripts/joy_cmd_vel.py
#!/usr/bin/env python
import rospy
import message_filters
from std_msgs.msg import Float32
from geometry_msgs.msg import Twist

def callback(yaw_rate, forward_vel,pub_vel):
	cmd_vel_msg = Twist()
    
	cmd_vel_msg.linear.x = forward_vel.data
	cmd_vel_msg.linear.y = 0
	cmd_vel_msg.linear.z = 0
	cmd_vel_msg.angular.z = -yaw_rate.data
	cmd_vel_msg.angular.x = 0
	cmd_vel_msg.angular.y = 0
	pub_vel.publish(cmd_vel_msg)

def joy_vel_task():
	rospy.init_node('joy_cmd_vel')	#definisco nome nodo
	pub_vel = rospy.Publisher('/robot/robotnik_base_control/cmd_vel', Twist, queue_size=1)	#definisco pubblicazione

	yaw_rate_sub_ = message_filters.Subscriber('unity/yaw_rate', Float32)
	forward_vel_sub_ = message_filters.Subscriber('unity/forward_vel', Float32)

	ts = message_filters.ApproximateTimeSynchronizer([yaw_rate_sub_, forward_vel_sub_], 10, 0.1, allow_headerless=True)
	ts.registerCallback(callback,pub_vel)
	rospy.spin()
	
if __name__ == '__main__':
	try:
		joy_vel_task()
	except rospy.ROSInterruptException:
		pass
